lib/IMAP/Threading/ThreadBuilder.php

```diff
@@ -0,0 +1,236 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * @copyright 2020 Christoph Wurst <christoph@winzerhof-wurst.at>
+ *
+ * @author 2020 Christoph Wurst <christoph@winzerhof-wurst.at>
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace OCA\Mail\IMAP\Threading;
+
+use OCA\Mail\Support\PerformanceLogger;
+use function array_key_exists;
+use function count;
+
+class ThreadBuilder {
+
+	/** @var PerformanceLogger */
+	private $performanceLogger;
+
+	public function __construct(PerformanceLogger $performanceLogger) {
+		$this->performanceLogger = $performanceLogger;
+	}
+
+	/**
+	 * @param Message[] $messages
+	 *
+	 * @return Container[]
+	 */
+	public function build(array $messages): array {
+		$log = $this->performanceLogger->start('Threading ' . count($messages) . ' messages');
+
+		// Step 1
+		$idTable = $this->buildIdTable($messages);
+		$log->step('build ID table');
+
+		// Step 2
+		$rootContainer = $this->buildRootContainer($idTable);
+		$log->step('build root container');
+
+		// Step 3
+		unset($idTable);
+		$log->step('free ID table');
+
+		// Step 4
+		$this->pruneContainers($rootContainer);
+		$log->step('prune containers');
+
+		// Step 5
+		$this->groupBySubject($rootContainer);
+		$log->step('group by subject');
+
+		$log->end();
+		// Return the children with reset numeric keys
+		return array_values($rootContainer->getChildren());
+	}
+
+	/**
+	 * @param Message[] $messages
+	 *
+	 * @return Container[]
+	 */
+	private function buildIdTable($messages): array {
+		/** @var Container[] $idTable */
+		$idTable = [];
+
+		foreach ($messages as $message) {
+			/** @var Message $message */
+			// Step 1.A
```

Comment:
Describe the steps for your future self?

Reply:
It's all at https://www.jwz.org/doc/threading.html, I didn't want to copy everything over